datumaro/plugins/voc_format/extractor.py

```diff
@@ -163,116 +190,150 @@ def __iter__(self):
-                    width = root_elem.find("size/width")
-                    if width is not None:
-                        width = int(width.text)
+                    if root_elem.getroot().tag != "annotation":
+                        raise MissingFieldError("annotation")
+
+                    height = self._parse_field(root_elem, "size/height", int, required=False)
+                    width = self._parse_field(root_elem, "size/width", int, required=False)
                     if height and width:
                         size = (height, width)
+
                     filename_elem = root_elem.find("filename")
                     if filename_elem is not None:
                         image = osp.join(image_dir, filename_elem.text)
-                    anns = self._parse_annotations(root_elem)
-                else:
+
+                    anns = self._parse_annotations(root_elem, item_id=(item_id, self._subset))
+
+                if image is None:
                     image = images.pop(item_id, None)
 
                 if image or size:
                     image = Image(path=image, size=size)
 
                 yield DatasetItem(id=item_id, subset=self._subset, media=image, annotations=anns)
+            except ElementTree.ParseError as e:
+                readable_wrapper = InvalidAnnotationError("Failed to parse XML file")
+                readable_wrapper.__cause__ = e
+                self._ctx.error_policy.report_item_error(
+                    readable_wrapper, item_id=(item_id, self._subset)
+                )
             except Exception as e:
-                self._report_item_error(e, item_id=(item_id, self._subset))
+                self._ctx.error_policy.report_item_error(e, item_id=(item_id, self._subset))
 
-    def _parse_annotations(self, root_elem):
-        item_annotations = []
-
-        for obj_id, object_elem in enumerate(root_elem.findall("object")):
-            obj_id += 1
-            attributes = {}
-            group = obj_id
-
-            obj_label_id = None
-            label_elem = object_elem.find("name")
-            if label_elem is not None:
-                obj_label_id = self._get_label_id(label_elem.text)
-
-            obj_bbox = self._parse_bbox(object_elem)
-
-            if obj_label_id is None or obj_bbox is None:
-                continue
-
-            difficult_elem = object_elem.find("difficult")
-            attributes["difficult"] = difficult_elem is not None and difficult_elem.text == "1"
-
-            truncated_elem = object_elem.find("truncated")
-            attributes["truncated"] = truncated_elem is not None and truncated_elem.text == "1"
-
-            occluded_elem = object_elem.find("occluded")
-            attributes["occluded"] = occluded_elem is not None and occluded_elem.text == "1"
+    @staticmethod
+    def _parse_field(
+        root, xpath: str, cls: Union[None, Type[T], Tuple[Type, ...]] = None, required: bool = True
+    ) -> Optional[Union[T, str]]:
```

Comment:
```suggestion
    @staticmethod
    def _parse_field(
        root: ElementTree.Element, xpath: str, cls: Type[T] = str, *, required: bool = True
    ) -> T:
```

Regarding the type of `cls`: the tuple alternative was not actually supported by the code; while the `None` alternative is unnecessary, as `str` gives the same result. You can prevent the construction of unnecessary strings by replacing `if cls` with `if cls is not str` in the function body.

Reply:
Changed